extensions/amp-analytics/0.1/transport.js

```diff
@@ -118,13 +124,194 @@ export class Transport {
+   * @param {!Object<string, string>} transportOptions
+   */
+  static sendRequestUsingCrossDomainIframe(request, transportOptions) {
+    const frameUrl = Transport.appendHashToUrl_(transportOptions['iframe'],
+      transportOptions['dataHash']);
+    const frameData = Transport.crossDomainFrames[frameUrl];
+    if (frameData.isReady) {
+      this.sendToCrossDomainIframe_(frameData.frame, [request]);
+    } else {
+      frameData.msgQueue.push(request);
+    }
+  }
+
+  /**
+   * Takes a URL and a data hash (the part that comes after the '#',
+   * optionally including the '#' itself) and concatenates them, adding the
+   * '#' if necessary
+   * @param {!string} url
+   * @param {string=} dataHash
+   * @return {!string}
+   * @private
+   */
+  static appendHashToUrl_(url, dataHash) {
+    if (dataHash) {
+      return url + (startsWith(dataHash, '#') ? dataHash : '#' + dataHash);
+    }
+    return url;
+  }
+
+  /**
+   * If iframe (and optionally dataHash as well) are specified in
+   * config/transport, check whether third-party iframe already exists, and if
+   * not, create it.
+   * @param {!HTMLDocument} ampDoc The AMP document
+   * @param {!Object<string,string>} transportOptions The 'transport' portion
+   * of the amp-analytics config object
+   * @param {Function} processResponse An optional function to receive any
+   * response messages back from the cross-domain iframe
+   */
+  static processCrossDomainIframe(ampDoc, transportOptions, processResponse) {
+    user().assert(!(transportOptions['beacon'] || transportOptions['xhrpost'] ||
+      transportOptions['image']), 'Cross-domain frame cannot coexist with' +
+      ' other transport methods');
+    const frameUrl = Transport.appendHashToUrl_(transportOptions['iframe'],
+        transportOptions['dataHash']);
+    // If iframe doesn't exist for this iframe url (and data hash), create it.
+    if (Transport.hasCrossDomainFrame(frameUrl)) {
+      this.incrementUsageCount_(frameUrl);
+    } else {
+      const frame = Transport.createCrossDomainFrame(ampDoc, frameUrl);
+      ampDoc.body.appendChild(frame);
+    }
+    if (processResponse) {
+      window.addEventListener("message", (msg) => {
+        if (msg.data.ampAnalyticsResponse) {
+          processResponse(msg.data.ampAnalyticsResponse);
+        }
+      });
+    }
+  }
+
+  /**
+   * Called when a creative no longer needs its cross-domain iframe (for
+   * instance, because the creative has been removed from the DOM).
+   * Once all creatives using a frame are done with it, the frame can be
+   * destroyed.
+   * @param {!HTMLDocument} ampDoc The AMP document
+   * @param {!Object<string,string>} transportOptions The 'transport' portion
+   * of the amp-analytics config object
+   */
+  static doneWithCrossDomainIframe(ampDoc, transportOptions) {
+    const frameUrl = Transport.appendHashToUrl_(transportOptions['iframe'],
+        transportOptions['dataHash']);
+    if (Transport.hasCrossDomainFrame(frameUrl)) {
```

Comment:
Combine predicates to a single if and I would argue that == 0 is sufficient but I suppose < can't hurt